components/common/src/main/java/com/hotels/styx/common/SequenceProcessor.java

```diff
@@ -0,0 +1,179 @@
+/*
+  Copyright (C) 2013-2019 Expedia Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+package com.hotels.styx.common;
+
+import reactor.core.publisher.Flux;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+
+/**
+ * Provides the ability to easily apply side effects and failure handling to the processing of a sequence of inputs
+ * without cluttering code with nested loops, try-catch and if-else statements.
+ *
+ * @param <T> input type
+ * @param <E> output type
+ */
+public class SequenceProcessor<T, E> {
```

Comment:
There are better ways to tackle our problem. Instead of implementing a bespoke sequence processor, you'd want go for monadic error handing, like `Try` or `Either` Scala types.

Consider implementing a simple `Try` class in Java. It won't be perfect but would have many advantages over a bespoke processor class: 
* It would work as a combinator with both java streams, reactor core flux, rx java and so on.
* More flexible as it is not limited to the operations supported by the sequence processor.
* Less code to maintain

For example:

    @Test
    public void testName() {
        List<Integer> inputs = asList(1, 2, 3, 4, 5, 6, 7);

        List<Try<String>> outcomes = inputs.stream()
                .map(number -> Try.tryBlock(() -> {
                    if (number % 2 == 0) {
                        throw new IllegalStateException();
                    }

                    return String.valueOf(number);
                }))
                .collect(Collectors.toList());

        List<String> successes = outcomes.stream()
                .filter(Try::isSuccess)
                .map(Try::result)
                .collect(Collectors.toList());

        List<Exception> errors = outcomes.stream()
                .filter(it -> !it.isSuccess())
                .map(Try::failure)
                .collect(Collectors.toList());
    }

    static class Try<T> {
        private final Exception exception;
        private final T result;

        Try(T result) {
            this.result = result;
            this.exception = null;
        }

        Try(Exception exception) {
            this.result = null;
            this.exception = exception;
        }

        public static <U, T> Try<T> tryBlock(Supplier<T> lambda) {
            try {
                return new Try<>(lambda.get());
            } catch (Exception e) {
                return new Try<>(e);
            }
        }

        public boolean isSuccess() {
            return this.exception == null;
        }

        public T result() {
            return this.result;
        }

        public Exception failure() {
            return this.exception;
        }
    }


With this the `SequenceProcessor` becomes obsolete.